numpy/core/src/umath/scalarmath.c.src

```diff
@@ -569,6 +571,112 @@ static void
+ * This function should only be used for scalar math fast paths.
+ *
+ * @param obj The object for which to find the descriptor
+ * @param descr The descriptor that was found.
+ * @returns 1 if the other object is a Python type and thus must not
+ *          handle the operation, otherwise 0. -1 _without_ an error set
+ *          if no descriptor was found.
+ */
+static int
+descr_from_basic_scalar(PyObject *obj, PyArray_Descr **descr)
+{
+    /* TODO: We could try giving defined scalars a chance... */
+    int type_num = PyArray_TypeNumFromNumPyScalarType(
+            (PyObject *)Py_TYPE(obj), 0);
+    if (type_num != NPY_NOTYPE) {
+        *descr = PyArray_DescrFromType(type_num);
+        return 0;
+    }
+    else if (PyFloat_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_DOUBLE);
+        return 1;
+    }
+    else if (PyBool_Check(obj)) {
+        *descr = PyArray_DescrFromType(NPY_BOOL);
+        return 1;
+    }
+    else if (PyLong_CheckExact(obj)) {
+        if (PyLong_AsLong(obj) == -1 && PyErr_Occurred()) {
+            PyErr_Clear();
+            return -1;
+        }
+        *descr = PyArray_DescrFromType(NPY_LONG);
+        return 1;
+    }
+    else if (PyComplex_CheckExact(obj)) {
+        *descr = PyArray_DescrFromType(NPY_CDOUBLE);
+        return 1;
+    }
+    return -1;
+}
+
+#define PyDescr_PythonCRepresentable(d) (d->type_num == NPY_LONGDOUBLE || \
+        d->type_num > NPY_CDOUBLE)
+
+/**
+ * This function attempts to compare NumPy or Python scalars
+ * The operation is done by getting value of both scalars
+ * and calling richcompare on them.
+ *
+ * @param self The first object which will be converted to scalar item
+ * @param other The second object to compare to
+ * @param cmp_op The value of comparison operator
+ * @returns Py_False or Py_True if richcompare is successfull
+ *          , otherwise NULL.
+ */
+static PyObject*
+do_richcompare_on_scalars(PyObject *self, PyObject *other, int cmp_op) {
+    PyObject *cmp_item_self, *cmp_item_other, *ret=NULL;
+    PyArray_Descr *self_descr=NULL, *other_descr=NULL;
+    void *data_self, *data_other;
+    int pyscalar_other, pyscalar_self;
+    int is_complex_operands, is_equality_operator, is_python_scalar;
+
+    pyscalar_self = descr_from_basic_scalar(self, &self_descr);
+    pyscalar_other = descr_from_basic_scalar(other, &other_descr);
+
+    if (pyscalar_self >= 0 && pyscalar_other >= 0) {
+        /*
+         * If either of the operands are complex and operator is not equality,
+         * or the operands are not representable as a native type,
+         * python's built-in richcompare cannot be used as it is not supported.
+         */
+        is_complex_operands = (PyTypeNum_ISCOMPLEX(self_descr->type_num) ||
+            PyTypeNum_ISCOMPLEX(other_descr->type_num));
+        is_equality_operator = (cmp_op != Py_EQ && cmp_op != Py_NE);
+        is_python_scalar = PyDescr_PythonCRepresentable(self_descr) ||
```

Comment:
I think a less confusing name would be `is_not_*` and `PyDescr_Not*`, will fix, aim is to basically check if the value can be represented as a native c type that python allows.

Reply:
No need for the "C" probably, we are aiming for converting (safely) to a native Python type?

Reply:
Yes indeed, also the `PyArray_Is*` functions that are already present do not have a proper check for native python types.

Reply:
The problem being that, if we cannot represent in a PyObject, the richcompare seems to cause problem, long double being one of the victims.